File: solutions/Spiral-Matrix-II/Spiral-Matrix-II.py

class Solution:
    # @return a list of lists of integer
    def generateMatrix(self, n):
        ans=[[[]for i in range(n)] for j in range(n)]
        start=0
        end=n-1
        num=1
        while start<end:
            for i in range(start,end):
                ans[start][i]=num
                num+=1
            for i in range(start,end):
                ans[i][end]=num
                num+=1
            for i in range(end,start,-1):
                ans[end][i]=num
                num+=1
            for i in range(end,start,-1):
                ans[i][start]=num
                num+=1
            start+=1
            end-=1
        if start==end:
            ans[start][start]=num
        return ans
 

File: solutions/Spiral-Matrix-II/Spiral-Matrix-II.js
/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function (n) {
  const ans = Array.from({ length: n }, (v, i) => {
    if (i === 0) {
      return Array.from({ length: n }, (v, j) => j + 1);
    } else {
      return Array.from({ length: n }, () => 0);
    }
  });
  if (n === 0) {
    return ans;
  }
  const dirs = [[1, 0], [0, -1], [-1, 0], [0, 1]];
  let dirIndex = 0;
  const pos = [0, n - 1];
  let m = n;
  let num = (m - 1) * n;
  const sum = n * m;
  while (num > 0) {
    for (let i = 1; i < m; i++) {
      num--;
      pos[0] += dirs[dirIndex][0];
      pos[1] += dirs[dirIndex][1];
      ans[pos[0]][pos[1]] = sum - num;
    }
    m--;
    [n, m] = [m, n];
    dirIndex = (dirIndex + 1) % dirs.length;
  }
  return ans;
};